internal/llrp/util_test.go

```diff
@@ -4,3 +4,317 @@
+		C1G2GetBlockPermalockStatusOpSpecResult *C1G2GetBlockPermalockStatusOpSpecResult
+		ClientRequestOpSpecResult               *ClientRequestOpSpecResult
+		Custom                                  []Custom
+	}
+
+	tests := []struct {
+		name     string
+		fields   fields
+		wantData string
+		wantOk   bool
+	}{
+		{
+			name:     "OK - nil",
+			fields:   fields{C1G2ReadOpSpecResult: nil},
+			wantData: "",
+			wantOk:   false,
+		},
+		{
+			name:     "OK - default values",
+			fields:   fields{C1G2ReadOpSpecResult: new(C1G2ReadOpSpecResult)},
+			wantData: wordsToHex([]uint16{}),
+			wantOk:   true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			rt := &TagReportData{
+				EPCData:                                 tt.fields.EPCData,
+				EPC96:                                   tt.fields.EPC96,
+				ROSpecID:                                tt.fields.ROSpecID,
+				SpecIndex:                               tt.fields.SpecIndex,
+				InventoryParameterSpecID:                tt.fields.InventoryParameterSpecID,
+				AntennaID:                               tt.fields.AntennaID,
+				PeakRSSI:                                tt.fields.PeakRSSI,
+				ChannelIndex:                            tt.fields.ChannelIndex,
+				FirstSeenUTC:                            tt.fields.FirstSeenUTC,
+				FirstSeenUptime:                         tt.fields.FirstSeenUptime,
+				LastSeenUTC:                             tt.fields.LastSeenUTC,
+				LastSeenUptime:                          tt.fields.LastSeenUptime,
+				TagSeenCount:                            tt.fields.TagSeenCount,
+				C1G2PC:                                  tt.fields.C1G2PC,
+				C1G2XPCW1:                               tt.fields.C1G2XPCW1,
+				C1G2XPCW2:                               tt.fields.C1G2XPCW2,
+				C1G2CRC:                                 tt.fields.C1G2CRC,
+				AccessSpecID:                            tt.fields.AccessSpecID,
+				C1G2ReadOpSpecResult:                    tt.fields.C1G2ReadOpSpecResult,
+				C1G2WriteOpSpecResult:                   tt.fields.C1G2WriteOpSpecResult,
+				C1G2KillOpSpecResult:                    tt.fields.C1G2KillOpSpecResult,
+				C1G2LockOpSpecResult:                    tt.fields.C1G2LockOpSpecResult,
+				C1G2BlockEraseOpSpecResult:              tt.fields.C1G2BlockEraseOpSpecResult,
+				C1G2BlockWriteOpSpecResult:              tt.fields.C1G2BlockWriteOpSpecResult,
+				C1G2RecommissionOpSpecResult:            tt.fields.C1G2RecommissionOpSpecResult,
+				C1G2BlockPermalockOpSpecResult:          tt.fields.C1G2BlockPermalockOpSpecResult,
+				C1G2GetBlockPermalockStatusOpSpecResult: tt.fields.C1G2GetBlockPermalockStatusOpSpecResult,
+				ClientRequestOpSpecResult:               tt.fields.ClientRequestOpSpecResult,
+				Custom:                                  tt.fields.Custom,
+			}
+			gotData, gotOk := rt.ReadDataAsHex()
+			require.Equal(t, gotData, tt.wantData)
+			require.Equal(t, gotOk, tt.wantOk)
```

Comment:
This is a case where I would use `require` first for `require.Equal(t, gotOk, tt.wantOk)` since if that fails, verifying the data isn't really valid. So something like:

```
require.Equal(t, gotOk, tt.wantOk)
assert.Equal(t, gotData, tt.wantData)
```

I.e. always first check the condition that indicates if the data is worth checking, like errors or in this case `ok`.